packages/lib/src/actionMiddlewares/draftMiddleware.ts

```diff
@@ -0,0 +1,102 @@
+import { getParentToChildPath, resolvePath } from "../parent"
+import { applyPatches, onPatches, OnPatchesDisposer, patchRecorder } from "../patch"
+import { isRootStore, registerRootStore, unregisterRootStore } from "../rootStore"
+import { clone } from "../snapshot"
+import { assertTweakedObject } from "../tweaker/core"
+
+/**
+ * Callback function for `DraftManager.withDraft`.
+ */
+export type WithDraftCallback<T extends object> = (draftNode: T) => boolean | undefined | void
+
+/**
+ * Manager class returned by `draftMiddleware` that allows you to make changes to a draft copy of
+ * the original subtree and apply them to the original subtree or reject them.
+ */
+export class DraftManager {
+  /**
+   * The draft copy of the original subtree.
+   */
+  private readonly subtreeRootDraft: object
+
+  /**
+   * The internal `onPatches` disposer.
+   */
+  private readonly disposer: OnPatchesDisposer
+
+  /**
+   * Creates an instance of `DraftManager`.
+   * Do not use directly, use `draftMiddleware` instead.
+   *
+   * @param subtreeRoot Subtree root target object.
+   */
+  public constructor(private readonly subtreeRoot: object) {
+    assertTweakedObject(subtreeRoot, "subtreeRoot")
+    this.subtreeRootDraft = clone(subtreeRoot)
+    if (isRootStore(subtreeRoot)) {
+      registerRootStore(this.subtreeRootDraft)
+    }
+    this.disposer = onPatches(subtreeRoot, patches => applyPatches(this.subtreeRootDraft, patches))
+  }
+
+  /**
+   * Executes `fn` with a draft of `node`. The changes made to the draft in `fn` can be accepted,
+   * i.e. applied to the original subtree, or rejected.
+   *
+   * @typeparam T Object type.
+   * @param node Object for which to obtain a draft copy.
+   * @param fn Function that is caled with a draft copy of `node`. Any changes made to the draft are
+   * applied to the original subtree when `fn` returns `true`. When `fn` returns `false`,
+   * `undefined` or nothing, the changes made to the draft are rejected.
+   */
+  public withDraft<T extends object>(node: T, fn: WithDraftCallback<T>): void {
+    assertTweakedObject(node, "node")
+
+    const path = getParentToChildPath(this.subtreeRoot, node)
+    if (!path) {
+      return
+    }
+
+    const draftNode = resolvePath<T>(this.subtreeRootDraft, path).value
+    if (!draftNode) {
+      return
+    }
+
+    const recorder = patchRecorder(this.subtreeRootDraft)
+    const accept = fn(draftNode)
+    recorder.dispose()
+
+    if (accept) {
+      applyPatches(
+        this.subtreeRoot,
+        recorder.events.flatMap(event => event.patches)
```

Comment:
Just wondering, how widespread is the support for flatmap nowadays?

Reply:
https://caniuse.com/#feat=array-flat

IE and Edge don't seem to support it plus a few other (mobile) browsers. After checking `mobx-keystone`'s other requirements like `WeakMap` etc., I think `flatMap` would reduce compatibility, so I'll use an alternative.

Reply:
Done in 7c67c7084d9b812f4468b441ee704ee6d63436c5. Do you prefer calling `applyPatches` multiple times (like it's done now) or materializing a new array and calling `applyPatches` only once?

There was a bug with regard to inverse patches order which I fixed in 91743704ebce5173fb7a6954adc2219cccc2df90.

Reply:
For performance reasons I usually prefer to avoid creating array copies whenever possible, so this approach looks good.